### PR TITLE
Added username and password arguments to pubsub.rabbitmq metadata

### PR DESCRIPTION
**Pull request by marioreggiori**
# Description

Added `username` and `password` arguments to pubsub.rabbitmq metadata and provide it to `amqp.Dial`.

**Caution:** This is probably a breaking change because the `host` argument was missused to also include credentials.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1719

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by berndverst** [comment]
How about when username and password are empty we fallback to the existing misused behavior to avoid the breaking change?

Could you add that please?

**Comment by marioreggiori** [comment]
We could omit username and password if the arguments haven't been provided via metadata. Unfortunately this is still a breaking change because we have to prepend the protocol `amqp://`.

**Comment by marioreggiori** [comment]
Hi @ItalyPaleAle, the idea was to split the connection URI into multiple arguments to support secrets generated by the rabbitmq-operator (containing username and password separately). It appears that referencing other variables with `$(PREVIOUSLY_DEFINED_VAR)` as possible with Kubernetes envs from secrets does not work within dapr `Components`.

**Comment by marioreggiori** [comment]
> I don't think there is any other protocol (like amqps) :)
Never mind.. I've should have read further :D

**Comment by berndverst** [comment]
Linter failure

```
  Error: amqp_authority contains underscore. You should use mixedCap or MixedCap. (nosnakecase)
  Error: amqp_authority contains underscore. You should use mixedCap or MixedCap. (nosnakecase)
  Error: amqp_authority contains underscore. You should use mixedCap or MixedCap. (nosnakecase)
```

**Comment by ItalyPaleAle** [comment]
@marioreggiori here's my proposal:

1. Create a new property "connectionString" which behaves like "host" does today.
1. Keep "host" as it is today and mark it as deprecated. It's an alias of "connectionString", taht will keep working but will not be in docs. If people use that, show a warning in the logs that they should change how they configure the component and use "connectionString" instead. Example of how we do that: https://github.com/dapr/dapr/blob/80f7e4ff9f478275d022cf2d0643028c36d43875/pkg/runtime/runtime.go#L2625-L2626
1. Create a new set of options: "hostname", "username", "password", "protocol". If there's no "connectionString", then construct it from reading from these.

**Comment by berndverst** [comment]
FYI our code freeze for Dapr 1.9 (when PRs must have been reviewed, reviews sufficiently addressed, and the PR merged) is a week from today.

This PR might not make this next release. The sooner it is completed, the better the chance.

**Comment by berndverst** [comment]
Please make sure to signoff on all of your commits too -- the DCO check is failing.

**Comment by marioreggiori** [comment]
@berndverst I'm having difficulties rewording the previous commit which I forgot to sign-off without having to resolve conflicts. Can you please give me a hint on how to solve this?

File: pubsub/rabbitmq/metadata.go [hunk 1]
```diff
@@ -76,10 +87,19 @@ func createMetadata(pubSubMetadata pubsub.Metadata) (*metadata, error) {
...
 
-	if val, found := pubSubMetadata.Properties[metadataHostKey]; found && val != "" {
-		result.host = val
-	} else {
-		return &result, fmt.Errorf("%s missing RabbitMQ host", errorMessagePrefix)
+	if val, found := pubSubMetadata.Properties[metadataConnectionStringKey]; found && val != "" {
+		result.connectionString = val
+	} else if val, found := pubSubMetadata.Properties[metadataHostKey]; found && val != "" {
+		result.connectionString = val
+		log.Warn("[DEPRECATION NOTICE] The 'host' argument is deprecated. Use 'connectionString' or individual connection arguments instead.")
```
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/metadata.go:94]: Can you add a link to the Dapr docs too in the log messages, which we'll update? https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-rabbitmq/

File: pubsub/rabbitmq/metadata.go [hunk 2]
```diff
@@ -184,3 +217,25 @@ func (m *metadata) formatQueueDeclareArgs(origin amqp.Table) amqp.Table {
...
+		if m.password != "" {
+			amqpAuthority.WriteString(":" + m.password)
+		}
+
+		amqpAuthority.WriteString("@")
+	}
+
+	amqpAuthority.WriteString(m.hostname)
+
+	return fmt.Sprintf("%s://%s", m.protocol, amqpAuthority.String())
```
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/metadata.go:240]: I also wonder if we should use `url` to build this, which would also take care of correctly escaping sequences when needed? WDYT?

Something like:

```go
		u := url.URL{}
		u.Scheme = m.protocol
		u.Host = m.hostname
		if m.username != "" {
			u.User = ...
		}
		return u.String()
```

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
LGTM, just a couple of small comments :)

**Comment by ItalyPaleAle** [comment]
>  I'm having difficulties rewording the previous commit which I forgot to sign-off without having to resolve conflicts. Can you please give me a hint on how to solve this?

Yeah, that's a bit of a pain in the bottom to solve. When I run into those situations, personally what I find easiest is to..... restart from scratch and re-apply the commits.

If you don't care about preserving the previous commits and are ok squashing them:

```sh
git reset --hard upstream/master
curl -L "https://patch-diff.githubusercontent.com/raw/dapr/components-contrib/pull/2129.diff" | git apply
```

Then commit with `git commit -s` as usual.

**Review by ItalyPaleAle — APPROVED** [review]

**Review by berndverst — APPROVED** [review]

**Comment by shubham1172** [comment]
@marioreggiori @berndverst this seems to be missing in the docs, is there a docs PR/issue to track this?

**Comment by berndverst** [comment]
I haven't seen anything. @marioreggiori did you open a PR against the docs repo? We reserve the right to revert any changes that are now documented.

**Comment by marioreggiori** [comment]
@ItalyPaleAle said that you will take care of the docs. See https://github.com/dapr/components-contrib/pull/2129#discussion_r979124785

> We reserve the right to revert any changes that are not documented.

I currently don't have time to work on this.. So feel free to revert.

**Comment by shubham1172** [comment]
Created https://github.com/dapr/docs/issues/3127 to track the docs changes.

**Pull request merged**